src/MSALWrapper/PCACache.cs

```diff
@@ -88,6 +100,153 @@ public void SetupTokenCache(ITokenCache userTokenCache, IList<Exception> errors)
+        /// <param name="errors">The errors list to append error encountered to.</param>
+        private void SetupPlainTextCache(ITokenCache userTokenCache, IList<Exception> errors)
+        {
+            try
+            {
+                var homeDir = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
+                var cacheDir = Path.Combine(homeDir, PlainTextCacheDir);
+                var cacheFilePath = Path.Combine(cacheDir, PlainTextCacheFileName);
+
+                // Create directory if it doesn't exist
+                if (!Directory.Exists(cacheDir))
+                {
+                    Directory.CreateDirectory(cacheDir);
+                    // Set directory permissions to user only (700)
+                    SetDirectoryPermissions(cacheDir);
+                }
+
+                // Create or ensure cache file exists with proper permissions
+                if (!File.Exists(cacheFilePath))
+                {
+                    File.WriteAllText(cacheFilePath, "{}");
+                    SetFilePermissions(cacheFilePath);
+                }
+                else
+                {
+                    // Ensure existing file has proper permissions
+                    SetFilePermissions(cacheFilePath);
+                }
+
+                var storageProperties = new StorageCreationPropertiesBuilder(PlainTextCacheFileName, cacheDir)
+                    .WithUnprotectedFile()
+                    .Build();
+
+                MsalCacheHelper cacher = MsalCacheHelper.CreateAsync(storageProperties).Result;
+                cacher.RegisterCache(userTokenCache);
+
+                this.logger.LogInformation($"Plain text cache fallback configured at: {cacheFilePath}");
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogWarning($"Plain text cache fallback failed: {ex.Message}");
+                errors.Add(ex);
+            }
+        }
+
+        /// <summary>
+        /// Checks if the current platform is Linux.
+        /// </summary>
+        /// <returns>True if running on Linux, false otherwise.</returns>
+        private static bool IsLinux()
+        {
+            return RuntimeInformation.IsOSPlatform(OSPlatform.Linux);
+        }
+
+        /// <summary>
+        /// Checks if the current Linux environment is headless (no display server).
+        /// </summary>
+        /// <returns>True if headless Linux environment, false otherwise.</returns>
+        private static bool IsHeadlessLinux()
+        {
+            // Check if DISPLAY environment variable is not set or empty
+            var display = Environment.GetEnvironmentVariable("DISPLAY");
+            if (string.IsNullOrEmpty(display))
+            {
+                return true;
+            }
+
+            // Check if WAYLAND_DISPLAY is not set or empty
+            var waylandDisplay = Environment.GetEnvironmentVariable("WAYLAND_DISPLAY");

```

Comment:
I might be reading the logic here wrong, but if I'm on an X11-only environment and `$DISPLAY` **is** set, but `$WAYLAND_DISPLAY` is _not_ set then this will report the environment as headless. I'm not sure if that's correct.